### PR TITLE
refactor: change byte prefixing scheme for provider

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: https://github.com/cosmos/interchain-security/issues/939

TODO:

- [ ] do the same on consumer

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Provided a link to the relevant issue or specification
- [ ] Reviewed "Files changed" and left comments if necessary 
- [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code 

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved handling of key prefixes to ensure correct retrieval and storage of data.

- **Refactor**
	- Updated key prefix logic across various functions for greater consistency and maintainability.
	- Refactored key management structure to enhance clarity and organization.

- **Tests**
	- Added new test functions to validate key prefix handling and ensure no prefix overlaps.
	- Enhanced existing tests for accuracy and better coverage of key prefix validations.

### CODE REVIEW

File: x/ccv/provider/types/keys.go [hunk 1]
```diff
@@ -33,231 +33,253 @@
...
-	MinimumPowerInTopNBytePrefix
+		// MinimumPowerInTopNKey is the key for storing the
+		// minimum power required to be in the top N per consumer chain.
+		"MinimumPowerInTopNKey": 40,
 
-	// NOTE: DO NOT ADD NEW BYTE PREFIXES HERE WITHOUT ADDING THEM TO getAllKeyPrefixes() IN keys_test.go
-)
+		// NOTE: DO NOT ADD NEW BYTE PREFIXES HERE WITHOUT ADDING THEM TO getAllKeyPrefixes() IN keys_test.go
+	}
+}
```
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/provider/types/keys.go]: This function is only used for testing. The non-determinism is not a problem.

File: x/ccv/provider/types/keys.go [hunk 2]
```diff
@@ -31,324 +32,499 @@
...
+// Only used for testing
+func GetAllKeyNames() []string {
+	prefixMap := getKeyPrefixes()
+	keys := make([]string, 0, len(prefixMap))
+	for k := range prefixMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply x/ccv/provider/types/keys.go:325]: This is only used for testing, as the comment above the function specifies.

File: x/ccv/provider/types/keys.go [hunk 3]
```diff
@@ -31,324 +32,499 @@
...
+	for k := range prefixMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	prefixList := make([]byte, 0, len(prefixMap))
+	for _, k := range keys {
+		prefixList = append(prefixList, prefixMap[k])
+	}
+	return prefixList
+}
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply x/ccv/provider/types/keys.go:313]: This is also only used for testing, where non-determinism is not an issue.

**Comment by p-offtermatt** [comment]
Just to double-check, this doesn't need a migration, since the actual keys remain the same, correct?

**Review by p-offtermatt — APPROVED** [review]
LGTM!

File: x/ccv/provider/types/keys.go [hunk 4]
```diff
@@ -31,324 +32,499 @@ const (
...
+	}
+}
+
+// mustGetKeyPrefix returns the key prefix for a given key.
+// It panics if there is not byte prefix for the index.
+func mustGetKeyPrefix(key string) byte {
+	keyPrefixes := getKeyPrefixes()
+	if prefix, found := keyPrefixes[key]; !found {
+		panic(fmt.Sprintf("could not find key prefix for index %s", key))
+	} else {
```
Comment (sainoe) [inline x/ccv/provider/types/keys.go:294]: very nit: `else` statement is unnecessary here.

**Comment by sainoe** [comment]
LGTM. Thanks for the refactor!

**Review by sainoe — APPROVED** [review]

File: x/ccv/provider/types/keys.go (hunk 4 above)
Comment (mpoke) [reply x/ccv/provider/types/keys.go:294]: without the else I need to define `prefix` outside the if.

**Pull request merged**